corehq/apps/data_analytics/tests/test_malt_generator.py

```diff
@@ -0,0 +1,130 @@
+        cls._setup_sofa_forms()
+        cls.run_malt_generation()
+
+    @classmethod
+    def tearDownClass(cls):
+        cls.domain.delete()
+
+    @classmethod
+    def _setup_domain_user(cls):
+        cls.domain = Domain(name=cls.DOMAIN_NAME)
+        cls.domain.save()
+        cls.user = CommCareUser.create(cls.DOMAIN_NAME, cls.USERNAME, '*****')
+        cls.user.save()
+        cls.user_id = cls.user._id
+
+    @classmethod
+    def _setup_apps(cls):
+        cls.app = Application.new_app(cls.DOMAIN_NAME, "app 1", APP_V2)
+        cls.wam_app = Application.new_app(cls.DOMAIN_NAME, "app 2", APP_V2)
+        cls.wam_app.amplifies_workers = 'yes'
+        cls.app.save()
+        cls.wam_app.save()
+        cls.app_id = cls.app._id
+        cls.wam_app_id = cls.wam_app._id
+
+    @classmethod
+    def _setup_sofa_forms(cls):
+        form_data_rows = []
+        common_args = {  # values don't matter
+            'time_start': cls.correct_date,
+            'time_end': cls.correct_date,
+            'duration': 10,
+        }
+
+        def _form_data(instance_id,
+                       app_id,
+                       domain=cls.DOMAIN_NAME,
+                       received_on=cls.correct_date,
+                       device_id=cls.DEVICE_ID,
+                       user_id=cls.user_id):
+            return FormData(
+                domain=domain,
+                received_on=received_on,
+                instance_id=instance_id,
+                device_id=device_id,
+                user_id=user_id,
+                app_id=app_id,
+                **common_args
+            )
+
+        def _append_forms(forms, received_on):
+            for form in forms:
+                instance_id, app_id = form
+                form_data_rows.append(
+                    _form_data(instance_id, app_id, received_on=received_on)
+                )
+
+        invalid_forms = [
+            ("out_of_range_1", cls.app_id),
+            ("out_of_range_2", cls.wam_app_id),
+            ("unknown_appId", cls.UNKNOWN_ID),
+        ]
+        valid_forms = [
+            ('app_form1', cls.app_id),
+            ('app_form2', cls.app_id),
+            ('app_form3', cls.app_id),
+            ('wam_app_form1', cls.wam_app_id),
+            ('wam_app_form2', cls.wam_app_id),
+            ('missing_app_form', MISSING_APP_ID),
+        ]
+
+        _append_forms(invalid_forms, cls.out_of_range_date)
+        _append_forms(valid_forms, cls.correct_date)
+
+        sms_form = _form_data('sms_form', cls.app_id, device_id=COMMCONNECT_DEVICE_ID)
+        form_data_rows.append(sms_form)
+
+        FormData.objects.bulk_create(form_data_rows)
+
+    @classmethod
+    def run_malt_generation(cls):
+        generator = MALTTableGenerator([cls.malt_month])
+        generator.build_table()
+
+    def test_per_user_per_app_counts(self):
+        def _assert_app_row_extists(app_id, num_forms=None, wam_value=None, count=1):
+            app_rows = MALTRow.objects.filter(
```

Comment:
There are ways to mock db tests, but in this case you really are verifying the way data is stored in the db, so I think db tests are quite appropriate.  It's a different story if you're testing business logic which only incidentally relies on the db.